interop-cats/shared/src/main/scala/zio/interop/cats.scala

```diff
@@ -35,348 +38,444 @@ abstract class CatsEffectPlatform
+    ZIO.interrupt.unit
+
+  override final def onCancel[A](fa: F[A], fin: F[Unit]): F[A] =
+    fa.onInterrupt(fin.ignore)
+
+  override final def memoize[A](fa: F[A]): F[F[A]] =
+    fa.memoize
+
+  override final def racePair[A, B](fa: F[A], fb: F[B]) =
+    (fa.interruptible raceWith fb.interruptible)(
+      (exit, fiber) => ZIO.succeedNow(Left((toOutcome(exit), toFiber(fiber)))),
+      (exit, fiber) => ZIO.succeedNow(Right((toFiber(fiber), toOutcome(exit))))
+    )
+
+  override final def both[A, B](fa: F[A], fb: F[B]): F[(A, B)] =
+    fa zipPar fb
+
+  override final def guarantee[A](fa: F[A], fin: F[Unit]): F[A] =
+    fa.ensuring(fin.orDieWith(fiberFailure))
+
+  override final def bracket[A, B](acquire: F[A])(use: A => F[B])(release: A => F[Unit]): F[B] =
+    acquire.bracket(release.andThen(_.orDieWith(fiberFailure)), use)
+
+  override val unique: F[Unique.Token] =
+    ZIO.effectTotal(new Unique.Token)
 }
 
-private class CatsConcurrent[R] extends CatsMonadError[R, Throwable] with Concurrent[RIO[R, *]] {
+private final class ZioDeferred[R, E, A](promise: Promise[E, A]) extends Deferred[ZIO[R, E, *], A] {
+  type F[T] = ZIO[R, E, T]
 
-  private[this] final def toFiber[A](f: Fiber[Throwable, A]): effect.Fiber[RIO[R, *], A] =
-    new effect.Fiber[RIO[R, *], A] {
-      override final val cancel: RIO[R, Unit] = f.interrupt.unit
-      override final val join: RIO[R, A]      = f.join
-    }
+  override val get: F[A] =
+    promise.await
 
-  override final def liftIO[A](ioa: effect.IO[A]): RIO[R, A] =
-    Concurrent.liftIO(ioa)(this)
+  override def complete(a: A): F[Boolean] =
+    promise.succeed(a)
 
-  override final def cancelable[A](k: (Either[Throwable, A] => Unit) => effect.CancelToken[RIO[R, *]]): RIO[R, A] =
-    ZIO.effectAsyncInterrupt { kk =>
-      val token = k(kk apply _.fold(ZIO.fail(_), ZIO.succeedNow))
-      Left(token.orDie)
+  override val tryGet: F[Option[A]] =
+    promise.isDone.flatMap {
+      case true  => get.asSome
+      case false => ZIO.none
     }
+}
 
-  override final def start[A](fa: RIO[R, A]): RIO[R, effect.Fiber[RIO[R, *], A]] =
-    fa.interruptible.forkDaemon.map(toFiber)
+private final class ZioRef[R, E, A](ref: ERef[E, A]) extends effect.Ref[ZIO[R, E, *], A] {
+  type F[T] = ZIO[R, E, T]
 
-  override final def racePair[A, B](
-    fa: RIO[R, A],
-    fb: RIO[R, B]
-  ): RIO[R, Either[(A, effect.Fiber[RIO[R, *], B]), (effect.Fiber[RIO[R, *], A], B)]] = {
-    def run[C](fc: RIO[R, C]): ZIO[R, Throwable, C] =
-      fc.interruptible.overrideForkScope(ZScope.global)
+  private def trySet(a: A): F[Boolean] =
+    set(a).as(true)
 
-    (run(fa) raceWith run(fb))(
-      { case (l, f) => l.fold(f.interrupt *> ZIO.halt(_), ZIO.succeedNow).map(lv => Left((lv, toFiber(f)))) },
-      { case (r, f) => r.fold(f.interrupt *> ZIO.halt(_), ZIO.succeedNow).map(rv => Right((toFiber(f), rv))) }
-    ).resetForkScope
-  }
+  override def access: F[(A, A => F[Boolean])] =
+    get.map(_ -> trySet)
```

Comment:
I think we need a more complicated state machine here because there are a couple of guarantees we have to honor. First, the `setter` is only supposed to set the value if the current value is still equal to the snapshot value returned by `access`. Second, the `setter` is supposed to only support being called once and return `false` on subsequent invocation. Something like:

```scala
  override def access: F[(A, A => F[Boolean])] =
    get.map { currentValue =>
      val called = new AtomicBoolean(false)
      def setter(a: A): F[Boolean] =
        ZIO.effectSuspendTotal {
          if (called.getAndSet(true)) {
            ZIO.succeedNow(false)
          } else {
            ref.modify { updatedValue =>
              if (currentValue == updatedValue) (true, a)
              else (false, updatedValue)
            }
          }
        }
      (currentValue, setter)
    }
```

Reply:
Good point - I guess the same applies to `tryUpdate`, `tryModify` and `tryModifyState`

Reply:
I think those are fine  because `update` and `modify` are atomic so they will always set the value based on a consistent snapshot of the initial value. We're just retrying under the hood until we succeed because we don't implement the very low level atomic operations at the effect system level. It is only in `access` where we are promising the ability to potentially do a set based on a consistent view of the state outside `modify`.